sys/kern/dtb.c

```diff
@@ -16,27 +16,114 @@ static uintptr_t _dtb_offset = 0;
 static void *_dtb_root = NULL;
 /* phys addr of dtb_root */
 static paddr_t _dtb_root_pa = 0;
-/* size of dtb */
+/* rounded size of dtb */
 static size_t _dtb_size = 0;
 
 paddr_t dtb_early_root(void) {
   assert(_dtb_root_pa != 0);
   return _dtb_root_pa + _dtb_offset;
 }
 
-void dtb_early_init(paddr_t dtb, size_t size) {
-  _dtb_root_pa = rounddown(dtb, PAGESIZE);
-  _dtb_offset = dtb - _dtb_root_pa;
-  _dtb_size = size + _dtb_offset;
+void dtb_early_init(paddr_t pa, vaddr_t va) {
+  void *dtb = (void *)va;
+
+  if (fdt_check_header(dtb))
+    panic("dtb incorrect header!");
+
+  size_t size = fdt_totalsize(dtb);
+
+  _dtb_root = (void *)va;
+  _dtb_root_pa = rounddown(pa, PAGESIZE);
+  _dtb_offset = pa - _dtb_root_pa;
+  _dtb_size = roundup(size + _dtb_offset, PAGESIZE);
 }
 
 void *dtb_root(void) {
   assert(_dtb_root != NULL);
   return _dtb_root + _dtb_offset;
 }
 
+size_t dtb_size(void) {
+  return _dtb_size;
+}
+
 void dtb_init(void) {
-  size_t dtb_size = roundup(_dtb_size, PAGESIZE);
   if (_dtb_root_pa != 0)
-    _dtb_root = (void *)kmem_map_contig(_dtb_root_pa, dtb_size, 0);
+    _dtb_root = (void *)kmem_map_contig(_dtb_root_pa, _dtb_size, 0);
+}
+
+static int dtb_offset(const char *path) {
+  int offset = fdt_path_offset(_dtb_root, path);
+  if (offset < 0)
+    panic("Failed to find dtb offset for %s!");
+  return offset;
+}
+
+static int dtb_addr_cells(int node) {
+  int len;
+  const uint32_t *prop = fdt_getprop(_dtb_root, node, "#address-cells", &len);
+  /* #address-cells is a 4-byte property */
+  if (prop == NULL || (size_t)len != sizeof(uint32_t))
+    panic("Failed to retreive #address-cells property for node %d!", node);
+  return fdt32_to_cpu(*prop);
+}
+
+static long dtb_to_cpu(const uint32_t *prop, int addr_cells) {
+  if (addr_cells == 1)
+    return fdt32_to_cpu(*prop);
+  assert(addr_cells == 2);
+  return fdt64_to_cpu(*(uint64_t *)prop);
+}
+
+/*
+ * XXX: FTTB, we consider only a single pair <addr size>.
+ */
+void dtb_reg(int parent, int node, uint64_t *addr_p, uint64_t *size_p) {
+  int len;
+  const int addr_cells = dtb_addr_cells(parent);
+  const uint32_t *prop = fdt_getprop(_dtb_root, node, "reg", &len);
+  if (prop == NULL || ((size_t)len % (2 * addr_cells * sizeof(uint32_t)) != 0))
```

Comment:
In this case it is a power of two and it would be better to check the bits with an and.
```suggestion
  if (prop == NULL || (((size_t)len & (2 * addr_cells * sizeof(uint32_t)) -1) != 0))
```

Reply:
You're right and we already have a macro for this purpose `is_aligned`. Fixed.